psi4/src/psi4/libfock/DFJCOSK.cc

```diff
@@ -0,0 +1,1012 @@
+    for (size_t bi = 0; bi < grid->blocks().size(); bi++) {
+
+        int rank = 0;
+#ifdef _OPENMP
+        rank = omp_get_thread_num();
+#endif
+
+        // grid points in this block
+        auto block = grid->blocks()[bi];
+        int npoints_block = block->npoints();
+        auto x = block->x();
+        auto y = block->y();
+        auto z = block->z();
+        auto w = block->w();
+
+        // significant basis functions and shells at these grid points
+        // significance determined via basis extent
+        const auto &bf_map = block->functions_local_to_global();
+        const auto &shell_map = block->shells_local_to_global();
+        int nbf_block = bf_map.size();
+        int ns_block = shell_map.size();
+
+        // lists of all basis functions and shells
+        //
+        // The use of these "all" lists adds O(N^2) cost to the COSK grid loop (w/ small prefactor)
+        // This cost is negligible relative to the esp integral computation, which is O(N) (w/ a much larger prefactor),
+        // but future COSK work could remove this potential bottleneck
+        std::vector<int> bf_map_all;
+        std::vector<int> shell_map_all;
+        for (size_t bf = 0; bf < nbf; bf++) bf_map_all.push_back(bf);
+        for (size_t s = 0; s < nshell; s++) shell_map_all.push_back(s);
+        int nbf_block_all = bf_map_all.size();
+        int ns_block_all = shell_map_all.size();
+
+        // => Bookkeeping <= //
+
+        // map index in shell_map_all to first index in bf_map_all
+        std::vector<int> shell_map_all_to_bf_map_all;
+        if (shell_map_all.size() > 0) {
+            shell_map_all_to_bf_map_all.push_back(0);
+            for(size_t shell_map_ind = 0; (shell_map_ind + 1) < shell_map_all.size(); shell_map_ind++) {
+                size_t MU = shell_map_all[shell_map_ind];
+                shell_map_all_to_bf_map_all.push_back(primary_->shell(MU).nfunction() + shell_map_all_to_bf_map_all.back());
+            }
+        }
+        
+        // map index in shell_map to first index in bf_map
+        std::vector<int> shell_map_to_bf_map;
+        if (shell_map.size() > 0) {
+            shell_map_to_bf_map.push_back(0);
+            for(size_t shell_map_ind = 0; (shell_map_ind + 1) < shell_map.size(); shell_map_ind++) {
+                size_t MU = shell_map[shell_map_ind];
+                shell_map_to_bf_map.push_back(primary_->shell(MU).nfunction() + shell_map_to_bf_map.back());
+            }
+        }
+
+        // map back from global shell index to local shell index
+        std::map<size_t, size_t> shell_map_inv;
+        for (size_t shell_ind = 0; shell_ind < shell_map.size(); shell_ind++) {
+            shell_map_inv[shell_map[shell_ind]] = shell_ind;
+        }
+
+        // => Process Density Matrix <= //
+
+        // significant cols of D for this grid block
+        std::vector<SharedMatrix> D_block(njk);
```

Comment:
Why is this `std::vector<SharedMatrix>` rather than `std::vector<Matrix>`?

Reply:
It wasn't obvious (to me at least) how to make this a `std::vector<Matrix>` without introducing extra copies.

For one, I had trouble initializing the vector with the `Matrix` constructor included. Any other way, and I think you'd have to initialize a `Matrix` and then separately push_back / copy it into the `std::vector`.

The other challenge is that the `linalg::doublet` function is later called with these matrices. This function has two signatures: either call it with two `SharedMatrix` objects and get a `SharedMatrix` back, or call it with two `Matrix` objects and get a `Matrix` back. The `F_block` variable that's created from the `doublet` call clearly needs to be a vector of `SharedMatrix`, so therefore `D_block` should also contain `SharedMatrix` objects.

Reply:
> For one, I had trouble initializing the vector with the `Matrix` constructor included. Any other way, and I think you'd have to initialize a `Matrix` and then separately push_back / copy it into the `std::vector`.

What's wrong with `D_block[jki] = Matrix(nbf_block_all, nbf_block);`?

> The `F_block` variable that's created from the `doublet` call clearly needs to be a vector of `SharedMatrix`

This isn't clear to me at all. Why can't it be `std::vector<Matrix>`?

Reply:
I was under the impression that the assignment:
```
D_block[jki] = Matrix(nbf_block_all, nbf_block);
```
should be avoided because it results in a redundant copy, same with the initialization of the elements of `F_block`

Reply:
You're right, per earlier comment. You _could_ use `emplace_back` instead of assigning the size of `D_block` in advance, but the extra resizes means that performance is [probably slightly better](https://stackoverflow.com/questions/65936284/performance-of-stdvectoremplace-back-vs-assignment-for-pod-struct) with the current version.

Oh, C++, why do you make things that should be simple so very technical.